### ISSUE TITLE
how to use a custom io take in fastapi

### ISSUE THREAD

**Issue by zexionghong** [tiangolo/fastapi#740]
**Description**

How can I [...]?

how to use a custom io take in fastapi?
I can't get the event_loop out......

**Comment by prostomarkeloff** [comment]
What do you want?

**Comment by zexionghong** [comment]
`import time
from fastapi import FastAPI
import asyncio

app = FastAPI()

def io():
    time.sleep(1)
    return None


@app.get('/')
async def index():
    # i want to use io() by async  ,but i cant
    io()
    return {‘xx’:''xx}
`
``

**Comment by zexionghong** [comment]
like this

**Comment by prostomarkeloff** [comment]
You can. Just define your route-function as synchronous and fastapi will run it in threadpool.

```py
import time
from fastapi import FastAPI
import asyncio

app = FastAPI()

def io():
time.sleep(1)
return None

@app.get('/')
def index():
io()
return {‘xx’:''xx}
```

**Comment by zexionghong** [comment]
for examples?thanks。。

**Comment by zexionghong** [comment]
can i get  await io()

**Comment by zexionghong** [comment]
Do not use async to define the view function with its own blocking task？

**Comment by zexionghong** [comment]
意思是有自己的阻塞任务的视图函数就不要使用async 定义函数么

**Comment by prostomarkeloff** [comment]
> Do not use async to define the view function with its own blocking task？

Yes. You should mark your functions with just `def` if you want to do inside blocking operations.

**Comment by zexionghong** [comment]
and i can write a ThreadPoolExecutor to run io()?

**Comment by prostomarkeloff** [comment]
> and i can write a ThreadPoolExecutor to run io()?

No, you just should write blocking code like this:

```py
# some definitions, etc.

@app.get("/")
def index():
    result = blocking_io()
    return {"result": result}

```

FastAPI will run your code in `ThreadPoolExecutor` by self.

**Comment by zexionghong** [comment]
Thanks

**Comment by sm-Fifteen** [comment]
You can also make your own custom async I/O functions, but those would have to either rely on async libraries (like httpx, aiohttp, aiofiles, asyncpg, etc.) or use Python asyncio's own "leaf futures" (the primitive IO functions those libraries are built upon) like `asyncio.sleep()`, `asyncio.open_connection()`, and so on.

```py
from fastapi import FastAPI
import asyncio

app = FastAPI()

async def io():
    await asyncio.sleep(1) # Don't forget to await your async I/O functions!
    return None

@app.get('/')
async def index():
    await io() # You have to await any function marked as `async def`
    return {'xx':'xx'}
```

**Comment by tiangolo** [comment]
Thanks for the help @prostomarkeloff and @sm-Fifteen :bowing_man: :cake:
